### DISCUSSION TITLE
Using Adsense in Next.js

### DISCUSSION THREAD

**Discussion by PolMrt** [vercel/next.js#18801] [Help] [answered]
Hey,

I'm currently trying to add adsense ads to my website. So here is what I've already done.

First, in `_document.js` I've added the Google Adsense script code : 

```js
import React from 'react'
import Document, { Html, Head, Main, NextScript } from 'next/document'

export default class MyDocument extends Document {
  render () {
    return (
      <Html>
        <Head>
          <script data-ad-client="ca-pub-****"
            type='text/javascript'
            async
            src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js" />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}

```

Secondly I've created a really simple `Adsense` component based on the npm package [react-adsense](https://www.npmjs.com/package/react-adsense) : 

```js
const Adsense = ({ client, slot, style, format = 'auto', layout = '', layoutKey = '', responsive = false, className = '' }) => {
  useEffect(() => {
    if (window) (window.adsbygoogle = window.adsbygoogle || []).push({})
  })

  return (
    <ins className={`${className} adsbygoogle`}
      style={style}
      data-ad-client={client}
      data-ad-slot={slot}
      data-ad-layout={layout}
      data-ad-layout-key={layoutKey}
      data-ad-format={format}
      data-full-width-responsive={responsive}></ins>
  )
}
```

However, this doesn't work. I get several errors : 
1. In my node terminal I get an error regarding the `window` : `UnhandledPromiseRejectionWarning: ReferenceError: window is not defined`. When I got this error for the first time I added the if statement in the useEffect but it doesn't work. I think this come from ssr page generation 🤔 
2. I get the error `Failed to load resource: the server responded with a status of 403 ()`in the browser console. I did some research about that but it seems to be normal since it's still in localhost and not in the registered domain of my adsense account.

So have you guys already implemented adsense in Next.js ? What am I doing wrong here ?

Thanks for your help !

**Comment by c0b41** [comment, marked as answer]
For head tags
```js
<script
   async
    src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"
 />
 <script
          dangerouslySetInnerHTML={{
            __html: `
               (adsbygoogle = window.adsbygoogle || []).push({
                   google_ad_client: "YOUR_ID",
                   enable_page_level_ads: true
              });
                `
          }}
 />
```

And for react components 
```js
useEffect(()=> {
      if(typeof window !== 'undefined'){
            (window.adsbygoogle = window.adsbygoogle || []).push({})
      }
}, [])
```

**Reply by PolMrt** [reply]
Thanks for your help 👌In fact, this works ! 

I was still getting the error `UnhandledPromiseRejectionWarning: ReferenceError: window is not defined` due to the npm package `just-detect-adblock` which also require to access `window`.

**Reply by hoonblizz** [reply]
I'm getting an error that says,
`Property 'adsbygoogle' does not exist on type 'Window & typeof globalThis'`
How did you manage to add adsbygoogle to window?
Thank you :)

**Reply by hoonblizz** [reply]
Sorry, resolved. 
For the future record, in the same file where we put `useEffect`,  we can do something like, 
```js
declare global {
  interface Window {
    adsbygoogle: {[key: string]: unknown}[]
  }
}

```

**Reply by cazterk** [reply]
repo?
